File: leetcode/Coin Change.js
// https://leetcode.com/problems/coin-change/

function coinChange(coins, amount) {
  const dp = new Array(amount + 1).fill(amount + 1);
  dp[0] = 0;
  for (const coin of coins) {
    for (let i = coin; i <= amount; ++i) {
      dp[i] = Math.min(dp[i], dp[i - coin] + 1);
    }
  }
  return dp[amount] != amount + 1 ? dp[amount] : -1;
}

let id = 1;

function test(coins, amount, expected) {
  const output = coinChange(coins, amount);
  const isCorrect = output === expected;
  console.log(`\nTest ${id++} ... ${isCorrect}`);
  if (!isCorrect) {
    console.log(`coins: ${coins}`);
    console.log(`amount: ${amount}`);
    console.error(`Output: ${output}`);
    console.error(`Expected: ${expected}`);
  }
}

test([1, 2, 5], 11, 3);
test([3], 3, 1);
test([2], 3, -1);
test([1], 0, 0);
test([186, 419, 83, 408], 6249, 20);
console.log("Finished.");

File: leetcode/Coin Change.py
# https://leetcode.com/problems/coin-change/
from typing import List
from tester import Tester


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        coins.sort()

        for coin in coins:
            for i in range(coin, amount + 1):
                dp[i] = min(dp[i], dp[i - coin] + 1)

        return dp[amount] if dp[amount] != amount + 1 else -1

    def coinChange(self, coins: List[int], amount: int) -> int:
        """https://leetcode.com/problems/coin-change/discuss/1545305/BFS-with-Bit-Manipulation"""
        step, seen = 0, [amount]
        checked = set()
        while seen:
            cur = []
            for node in seen:
                if node == 0:
                    return step
                for coin in coins:
                    rem = node - coin
                    if rem >= 0 and rem not in checked:
                        cur.append(rem)
                        checked.add(rem)
            step, seen = step + 1, cur
        return -1

    def coinChange(self, coins: List[int], amount: int) -> int:
        """Bitwise BFS."""
        step, seen = 0, 1 << amount
        while seen & 1 != 1:
            cur = seen
            for coin in coins:
                cur |= seen >> coin
            if cur == seen:
                return -1
            step, seen = step + 1, cur
        return step


t = Tester(Solution())

t.test(3, [1, 2, 5], 11)
t.test(-1, [2], 3)
t.test(0, [1], 0)

t.report()
